File: lc_soln/0098_valid_binary_search_tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, INT_MIN, INT_MAX);
    }

    bool dfs(TreeNode* root, int minVal, int maxVal) {
        if (root == nullptr) {
            return true;
        }
        if (!(minVal < root->val && root->val < maxVal)) {
            return false;
        }
        return dfs(root->left, minVal, root->val) && dfs(root->right, root->val, maxVal);
    }
};

int main() {
    return 0;
}

File: lc_soln/0098_valid_binary_search_tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution
{
    public bool IsValidBST(TreeNode root)
    {
        return Dfs(root, int.MinValue, int.MaxValue);
    }

    bool Dfs(TreeNode root, int minVal, int maxVal)
    {
        if (root is null)
        {
            return true;
        }
        if (!(minVal < root.val && root.val < maxVal))
        {
            return false;
        }
        return Dfs(root.left, minVal, root.val) && Dfs(root.right, root.val, maxVal);
    }
}
